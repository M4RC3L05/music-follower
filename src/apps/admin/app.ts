import config from "config";
import Koa from "koa";
import bodyParser from "koa-bodyparser";
import render from "koa-ejs";
import qs from "koa-qs";
import session from "koa-session";
import koaStatic from "koa-static";
import path from "node:path";
import url from "node:url";

import router from "#src/apps/admin/router.js";

export const app = () => {
  const koa = new Koa();
  koa.keys = config.get("apps.admin.keys");

  qs(koa);
  render(koa, {
    root: path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), "views"),
    layout: "main",
    viewExt: "ejs",
    cache: process.env.NODE_ENV === "production",
    debug: false,
  });

  koa.use(session({ ...config.get("apps.admin.session") }, koa));
  koa.use(bodyParser());
  koa.use(koaStatic(path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), "../../../public")));
  koa.use(router.middleware());
  koa.use(router.allowedMethods());

  return koa;
};
