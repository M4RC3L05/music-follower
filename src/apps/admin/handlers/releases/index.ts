import { zValidator } from "@hono/zod-validator";
import type { Hono } from "hono";
import { z } from "zod";
import { RequestValidationError } from "#src/errors/mod.js";
import { utils } from "../../views/common/mod.js";
import { releasesViews } from "../../views/mod.js";

export const handler = (router: Hono) => {
  router.get(
    "/releases",
    zValidator(
      "query",
      z
        .object({
          q: z.string().optional(),
          hidden: z.enum(["feed", "admin", ""]).optional(),
          page: z.string().regex(/^\d+$/).optional(),
          limit: z.string().regex(/^\d+$/).optional(),
        })
        .strict(),
      (result) => {
        if (!result.success)
          throw new RequestValidationError({
            request: { query: result.error },
          });
      },
    ),
    async (c) => {
      const { hidden, limit, page, q } = c.req.valid("query");
      const { data: releases, pagination } = await c
        .get("services")
        .api.releasesService.getReleases({
          query: {
            hidden: hidden === "" ? undefined : hidden,
            limit,
            page,
            q,
            notHidden: !hidden || hidden.length <= 0 ? "admin" : undefined,
          },
        });

      const previousLink = `/releases?page=${pagination.previous}${
        pagination.limit ? `&limit=${pagination.limit}` : ""
      }${q ? `&q=${q}` : ""}${hidden ? `&hidden=${hidden}` : ""}`;
      const nextLink = `/releases?page=${pagination.next}${
        pagination.limit ? `&limit=${pagination.limit}` : ""
      }${q ? `&q=${q}` : ""}${hidden ? `&hidden=${hidden}` : ""}`;

      return c.html(
        releasesViews.pages.Index({
          releases,
          pagination: {
            currentUrl: c.req.url,
            startLink: utils.replaceAndReloadLink(
              previousLink.replace(/\?page=[0-9]+/, "?page=0"),
            ),
            previousLink: utils.replaceAndReloadLink(previousLink),
            nextLink: utils.replaceAndReloadLink(nextLink),
            endLink: utils.replaceAndReloadLink(
              previousLink.replace(
                /\?page=[0-9]+/,
                `?page=${Math.floor(pagination.total / pagination.limit)}`,
              ),
            ),
          },
        }),
      );
    },
  );
};
