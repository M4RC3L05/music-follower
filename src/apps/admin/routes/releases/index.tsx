import type { Hono } from "@hono/hono";
import { ReleasesIndexPage } from "#src/apps/admin/views/releases/pages/index.tsx";

export const index = (router: Hono) => {
  router.get("/", async (c) => {
    const { hidden, limit, page, q } = c.req.query();
    const { data: releases, pagination } = await c
      .get("services")
      .api.releasesService.getReleases({
        hidden: hidden === "" ? undefined : hidden,
        limit: limit ? Number(limit) : undefined,
        page: page ? Number(page) : undefined,
        q,
        notHidden: !hidden || hidden.length <= 0 ? "admin" : undefined,
        signal: c.req.raw.signal,
      });

    const previousLink = `/releases?page=${pagination.previous}${
      pagination.limit ? `&limit=${pagination.limit}` : ""
    }${q ? `&q=${q}` : ""}${hidden ? `&hidden=${hidden}` : ""}`;
    const nextLink = `/releases?page=${pagination.next}${
      pagination.limit ? `&limit=${pagination.limit}` : ""
    }${q ? `&q=${q}` : ""}${hidden ? `&hidden=${hidden}` : ""}`;

    return c.render(
      <ReleasesIndexPage
        releases={releases}
        pagination={{
          currentUrl: c.req.url,
          startLink: previousLink.replace(/\?page=[0-9]+/, "?page=0"),
          previousLink: previousLink,
          nextLink: nextLink,
          endLink: previousLink.replace(
            /\?page=[0-9]+/,
            `?page=${Math.floor(pagination.total / pagination.limit)}`,
          ),
        }}
      />,
    );
  });
};
