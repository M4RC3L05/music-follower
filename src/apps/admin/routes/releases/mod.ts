import { Hono } from "hono";
import { index } from "#src/apps/admin/routes/releases/index.tsx";
import { show } from "#src/apps/admin/routes/releases/show.tsx";
import { update } from "#src/apps/admin/routes/releases/update.ts";

export const releasesRoutes = () => {
  const router = new Hono();

  index(router);
  show(router);
  update(router);

  return router;
};
