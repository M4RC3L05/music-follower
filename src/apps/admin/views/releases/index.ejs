<div class="container-xl mt-4">
  <div class="row">
    <h1>Releases</h1>
  </div>

  <div class="row">
    <form method="get" action="/releases?<%= query ? `q=${query}` : "" %>">
      <div class="mb-3">
        <input type="search" placeholder="Search releases..." name="q" class="form-control" value="<%= query %>">
      </div>
    </form>
  </div>

  <div class="row row-cols-1 row-cols-md-3 row-cols-sm-2 g-4 mb-4">
    <% releases.forEach((release)=> { %>
    <div class="col">
      <div class="card text-bg-dark">
        <a href="/releases/<%= release.id %>?type=<%= release.type %>"
          style="position: absolute; top: 8px; right: 8px; z-index: 2;" class="btn btn-primary"><i
            class="bi bi-info-circle-fill"></i></a>

        <img src="<%= release.coverUrl %>" class="card-img" style="aspect-ratio: 1 / 1;">
        <div class="card-img-overlay"
          style="display: flex; justify-content: end; flex-direction: column; background: linear-gradient(-180deg,transparent 20%,rgba(0,0,0,.7) 100%)">
          <div class="mb-2">
            <% if (release.releasedAt.getTime()> Date.now()) { %>
            <span class="badge text-bg-info">To be released</span>
            <% } %>
            <span class="badge text-bg-success">
              <%= release.type %>
            </span>
          </div>
          <h5 class="card-title"><strong>
              <%= release.name %> by <%= release.artistName %>
            </strong></h5>
          <p class="card-text">Released at <%= release.releasedAt.toLocaleString() %>
          </p>
        </div>
      </div>
    </div>
    <% }) %>
  </div>

  <div class="row">
    <div class="col">
      <nav>
        <ul class="pagination justify-content-center">
          <li class="page-item <%= page <= 0 ? " disabled" : "" %>"><a class="page-link"
              href="/releases?page=0<%= query ? `&q=${query}` : "" %>">Start</a></li>
          <li class="page-item <%= page <= 0 ? " disabled" : "" %>"><a class="page-link"
              href="/releases?page=<%= Math.max(page - 1, 0) %><%= query ? `&q=${query}` : "" %>">Previous</a></li>
          <li class="page-item <%= (page + 1) * limit >= total - 1 ? " disabled" : "" %>"><a class="page-link"
              href="/releases?page=<%= page + 1 %><%= query ? `&q=${query}` : "" %>">Next</a></li>
          <li class="page-item <%= (page + 1) * limit >= total - 1 ? " disabled" : "" %>"><a class="page-link"
              href="/releases?page=<%= Math.ceil(total / limit) - 1 %><%= query ? `&q=${query}` : "" %>">End</a></li>
        </ul>
      </nav>
    </div>
  </div>
</div>
