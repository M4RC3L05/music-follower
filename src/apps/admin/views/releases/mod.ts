export * as pages from "#src/apps/admin/views/releases/pages/mod.ts";
