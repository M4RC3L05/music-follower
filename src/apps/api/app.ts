import { type Database } from "@leafac/sqlite";
import config from "config";
import { Hono } from "hono";
import { basicAuth } from "hono/basic-auth";
import { cors } from "hono/cors";
import { errorMappers } from "#src/errors/mod.js";
import { errorMapper, requestLifeCycle } from "#src/middlewares/mod.js";
import { router } from "./router.js";

export type Api = ReturnType<typeof makeApp>;

export const makeApp = ({ database }: { database: Database }) => {
  const app = new Hono();

  app.use("*", (c, next) => {
    c.set("database", database);

    return next();
  });
  app.use("*", requestLifeCycle);
  app.use("*", cors());
  app.use(
    "*",
    basicAuth({
      username: config.get<{ name: string; pass: string }>("apps.api.basicAuth")
        .name,
      password: config.get<{ name: string; pass: string }>("apps.api.basicAuth")
        .pass,
    }),
  );

  app.onError(
    errorMapper({
      mappers: [errorMappers.validationErrorMapper],
      defaultMapper: errorMappers.defaultErrorMapper,
    }),
  );

  return app.route("/api", router());
};
