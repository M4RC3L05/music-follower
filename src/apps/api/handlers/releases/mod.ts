import { Env, Hono } from "hono";
import { SchemaType } from "#src/common/utils/types.js";
import { default as getRelease } from "./get.js";
import { default as getReleases } from "./search.js";
import { default as updateRelease } from "./update.js";

export const router = () => {
  let router = new Hono();

  router = getRelease(router);
  router = getReleases(router);
  router = updateRelease(router);

  return router as Hono<
    Env,
    SchemaType<ReturnType<typeof getRelease>> &
      SchemaType<ReturnType<typeof getReleases>> &
      SchemaType<ReturnType<typeof updateRelease>>,
    "/"
  >;
};
