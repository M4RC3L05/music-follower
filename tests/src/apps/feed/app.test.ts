import { createServer } from "node:http";

import { describe, expect, jest, test } from "@jest/globals";
import config from "config";
import { makeFetch } from "supertest-fetch";

import { app } from "#src/apps/feed/app.js";
import { loadRelease } from "#tests/fixtures/releases.js";

const fetch = makeFetch(createServer(app().callback()));

describe("app", () => {
  test("it should throw 400 if it did not receive the format query", async () => {
    const data = await fetch("/");

    expect(data.status).toBe(400);
    expect(await data.text()).toBe("Invalid feed format provided");
  });

  test("it should throw 400 if the format provided is invalid", async () => {
    const data = await fetch("/?format=foo");

    expect(data.status).toBe(400);
    expect(await data.text()).toBe("Invalid feed format provided");
  });

  test('it should provide a feed in the format "rss"', async () => {
    await loadRelease({
      feedAt: new Date(2),
      id: 1,
      artistName: "foo",
      coverUrl: "http://foo.local/foo.png",
      name: "fiz",
      releasedAt: new Date(0),
      type: "collection",
    });
    await loadRelease({
      feedAt: new Date(0),
      id: 2,
      artistName: "foo",
      coverUrl: "http://foo.local/foo.png",
      name: "fiz",
      releasedAt: new Date(1000),
      type: "track",
    });
    await loadRelease({
      feedAt: new Date(1),
      id: 3,
      artistName: "foo",
      coverUrl: "http://foo.local/foo.png",
      name: "fiz",
      releasedAt: new Date(2000),
      type: "collection",
    });

    jest.spyOn(config, "get").mockReturnValueOnce(2);
    jest.useFakeTimers().setSystemTime(0);
    const data = await fetch("/?format=rss").expectStatus(200);

    expect(data.headers.get("content-type")).toBe("application/rss+xml");
    expect(await data.text()).toMatchSnapshot();
  });

  test('it should provide a feed in the format "atom"', async () => {
    await loadRelease({
      feedAt: new Date(2),
      id: 1,
      artistName: "foo",
      coverUrl: "http://foo.local/foo.png",
      name: "fiz",
      releasedAt: new Date(0),
      type: "collection",
    });
    await loadRelease({
      feedAt: new Date(0),
      id: 2,
      artistName: "foo",
      coverUrl: "http://foo.local/foo.png",
      name: "fiz",
      releasedAt: new Date(1000),
      type: "track",
    });
    await loadRelease({
      feedAt: new Date(1),
      id: 3,
      artistName: "foo",
      coverUrl: "http://foo.local/foo.png",
      name: "fiz",
      releasedAt: new Date(2000),
      type: "collection",
    });

    jest.spyOn(config, "get").mockReturnValueOnce(2);
    jest.useFakeTimers().setSystemTime(0);
    const data = await fetch("/?format=atom").expectStatus(200);

    expect(data.headers.get("content-type")).toBe("application/atom+xml");
    expect(await data.text()).toMatchSnapshot();
  });

  test('it should provide a feed in the format "json"', async () => {
    await loadRelease({
      feedAt: new Date(2),
      id: 1,
      artistName: "foo",
      coverUrl: "http://foo.local/foo.png",
      name: "fiz",
      releasedAt: new Date(0),
      type: "collection",
    });
    await loadRelease({
      feedAt: new Date(0),
      id: 2,
      artistName: "foo",
      coverUrl: "http://foo.local/foo.png",
      name: "fiz",
      releasedAt: new Date(1000),
      type: "track",
    });
    await loadRelease({
      feedAt: new Date(1),
      id: 3,
      artistName: "foo",
      coverUrl: "http://foo.local/foo.png",
      name: "fiz",
      releasedAt: new Date(2000),
      type: "collection",
    });

    jest.spyOn(config, "get").mockReturnValueOnce(2);
    jest.useFakeTimers().setSystemTime(0);
    const data = await fetch("/?format=json").expectStatus(200);

    expect(data.headers.get("content-type")).toBe("application/json; charset=utf-8");
    expect(await data.text()).toMatchSnapshot();
  });
});
