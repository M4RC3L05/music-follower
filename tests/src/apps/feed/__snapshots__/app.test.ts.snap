// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app it should provide a feed in the format "atom" 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">
    <id>music_follower</id>
    <title>Music releases</title>
    <updated>1970-01-01T00:00:00.000Z</updated>
    <generator>Music Follower</generator>
    <subtitle>Get the latest music releases from artist you follow</subtitle>
    <rights>Music Follower</rights>
    <entry>
        <title type="html"><![CDATA[fiz by foo]]></title>
        <id>1</id>
        <link href="http://foo.local/foo.png"/>
        <updated>1970-01-01T00:00:00.002Z</updated>
        <summary type="html"><![CDATA[<img src="http://foo.local/foo.png" />
      <h1>fiz</h1><p>foo</p>
      <p>1970</p>]]></summary>
    </entry>
    <entry>
        <title type="html"><![CDATA[fiz by foo]]></title>
        <id>3</id>
        <link href="http://foo.local/foo.png"/>
        <updated>1970-01-01T00:00:00.001Z</updated>
        <summary type="html"><![CDATA[<img src="http://foo.local/foo.png" />
      <h1>fiz</h1><p>foo</p>
      <p>1970</p>]]></summary>
    </entry>
</feed>"
`;

exports[`app it should provide a feed in the format "json" 1`] = `
"{
    "version": "https://jsonfeed.org/version/1",
    "title": "Music releases",
    "description": "Get the latest music releases from artist you follow",
    "items": [
        {
            "id": "1",
            "url": "http://foo.local/foo.png",
            "title": "fiz by foo",
            "summary": "<img src=\\"http://foo.local/foo.png\\" />\\n      <h1>fiz</h1><p>foo</p>\\n      <p>1970</p>",
            "image": "http://foo.local/foo.png",
            "date_modified": "1970-01-01T00:00:00.002Z"
        },
        {
            "id": "3",
            "url": "http://foo.local/foo.png",
            "title": "fiz by foo",
            "summary": "<img src=\\"http://foo.local/foo.png\\" />\\n      <h1>fiz</h1><p>foo</p>\\n      <p>1970</p>",
            "image": "http://foo.local/foo.png",
            "date_modified": "1970-01-01T00:00:00.001Z"
        }
    ]
}"
`;

exports[`app it should provide a feed in the format "rss" 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<rss version="2.0">
    <channel>
        <title>Music releases</title>
        <link>undefined</link>
        <description>Get the latest music releases from artist you follow</description>
        <lastBuildDate>Thu, 01 Jan 1970 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>Music Follower</generator>
        <language>en</language>
        <copyright>Music Follower</copyright>
        <item>
            <title><![CDATA[fiz by foo]]></title>
            <link>http://foo.local/foo.png</link>
            <guid>1</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:00 GMT</pubDate>
            <description><![CDATA[<img src="http://foo.local/foo.png" />
      <h1>fiz</h1><p>foo</p>
      <p>1970</p>]]></description>
            <enclosure url="http://foo.local/foo.png" length="0" type="image/png"/>
        </item>
        <item>
            <title><![CDATA[fiz by foo]]></title>
            <link>http://foo.local/foo.png</link>
            <guid>3</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:00 GMT</pubDate>
            <description><![CDATA[<img src="http://foo.local/foo.png" />
      <h1>fiz</h1><p>foo</p>
      <p>1970</p>]]></description>
            <enclosure url="http://foo.local/foo.png" length="0" type="image/png"/>
        </item>
    </channel>
</rss>"
`;
